pkg/services/store/object/object.proto

```diff
@@ -0,0 +1,242 @@
+  string previous_version = 6;
+}
+
+message WriteObjectResponse {
+  // Error info -- if exists, the save did not happen
+  ObjectErrorInfo error = 1;
+
+  // Object details with the body removed
+  RawObject object = 2; 
+
+  // Object summary as JSON
+  bytes summary_json = 3;
+}
+
+//-----------------------------------------------
+// Delete request/response
+//-----------------------------------------------
+
+message DeleteObjectRequest {
+  string UID = 1;
+  string kind = 2;
+
+  // Used for optimistic locking.  If missing, the previous version will be replaced regardless
+  string previous_version = 3;
+}
+
+message DeleteObjectResponse {
+  bool OK = 1;
+}
+
+//-----------------------------------------------
+// History request/response
+//-----------------------------------------------
+
+message ObjectHistoryRequest {
+  // Object identifier
+  string UID = 1;
+
+  // type of object we are listing
+  string kind = 2;
+
+  // Maximum number of items to return
+  int64 limit = 3; 
+
+  // Starting from the requested page
+  string next_page_token = 5;
+}
+
+message ObjectHistoryResponse {
+  // Object metadata without the raw bytes
+  repeated RawObject object = 1;
+
+  // All users who have edited the object
+  repeated UserInfo authors = 2;
+
+  // The object creation time
+  int64 created_time = 3; 
+
+  // More results exist... pass this in the next request
+  string next_page_token = 4;
+}
+
+
+//-----------------------------------------------
+// List request/response
+//-----------------------------------------------
+
+message ObjectSearchRequest {
+  // Starting from the requested page (other query parameters must match!)
+  string next_page_token = 1;
+
+  // Maximum number of items to return
+  int64 limit = 2; 
+
+  // Free text query string -- mileage may vary :)
+  string query = 3;
+
+  // limit to a specific kind (empty is all)
+  repeated string kind = 4;
+
+  // Limit results to items in a specific folder
+  string folder = 5;
+
+  // Must match all labels
+  map<string,string> labels = 6;
+
+  // Find items that match a given reference
+  ExternalReference reference = 7;
+
+  // Sorting instructions `field ASC/DESC`
+  repeated string sort = 8;
+
+  // TODO, limit the set of fields we actually want returned
+  // Only supported in the QueryResponse flavor?
+  repeated string fields = 9;
+
+  // Return the full body in each payload
+  bool with_body = 10;
+}
+
+message ObjectSearchResponse {
+  repeated RawObject results = 1;
+
+  // More results exist... pass this in the next request
+  string next_page_token = 2;
+}
+
+
+//-----------------------------------------------
+// Storage interface
+//-----------------------------------------------
+
+// This assumes a future grpc interface where the user info is passed in context, not in each message body
+// for now it will only work with an admin API key
+service ObjectStore {
+  rpc Read(ReadObjectRequest) returns (ReadObjectResponse);
```

Comment:
same as in the plugin-sdk PR, I think we need a batch https://github.com/grafana/grafana-plugin-sdk-go/pull/527#discussion_r963660328

Reply:
yes, was about to say the same thing a batch read will be essential (potentially a batch insert/update as well but think we can live without that longer)

Reply:
Do we want/need an independent "batch" call, if search has a "with_body" option?  https://github.com/grafana/grafana/pull/55646/files#diff-675aa0519d14d16218009e25aa53ea20f3965343829666ab87133185b1e2198fR237

Reply:
I guess we could include it in search, but currently the search request supports only a single reference:

```
  // Find items that match a given reference
  ExternalReference reference = 7;
```

also I am not exactly sure what does the comment mean - would it match only on `object.uid` or on `object.references.uid`?

Reply:
changed it to repeated value and added a better comment :)

Reply:
> Do we want/need an independent "batch" call, if search has a "with_body" option?

I think so, just for ease of use / discovery. Afraid many will miss it otherwise and do get's in a loop :)

Reply:
added